[Build] Ubuntu 16 CUDA

<!--
A short one or two line description of the error
-->

Description
===========
Hi. Could you look at the error and tell what I am doing wrong?
<!--
* Additional details about the errors during the build
* What do you suspect is causing the issue?
* Which steps in the (wiki)[https://github.com/arrayfire/arrayfire/wiki] failed
* What operating system and/or distro are you using?
* Versions of the packages related to this bug
-->

Error Log
---------
<!-- The output of the error log. -->
```
turgut@turgut:~$ cd ~/Documents/arrayfire
turgut@turgut:~/Documents/arrayfire$ mkdir build && cd build
turgut@turgut:~/Documents/arrayfire/build$ cmake .. -DCMAKE_BUILD_TYPE=Release -DAF_BUILD_CUDA=ON -DAF_BUILD_CPU=OFF -DAF_BUILD_OPENCL=OFF
-- The C compiler identification is GNU 5.4.0
-- The CXX compiler identification is GNU 5.4.0
-- Check for working C compiler: /usr/bin/cc
-- Check for working C compiler: /usr/bin/cc -- works
-- Detecting C compiler ABI info
-- Detecting C compiler ABI info - done
-- Detecting C compile features
-- Detecting C compile features - done
-- Check for working CXX compiler: /usr/bin/c++
-- Check for working CXX compiler: /usr/bin/c++ -- works
-- Detecting CXX compiler ABI info
-- Detecting CXX compiler ABI info - done
-- Detecting CXX compile features
-- Detecting CXX compile features - done
-- Looking for CL_VERSION_2_0
-- Looking for CL_VERSION_2_0 - not found
-- Looking for CL_VERSION_1_2
-- Looking for CL_VERSION_1_2 - not found
-- Looking for CL_VERSION_1_1
-- Looking for CL_VERSION_1_1 - not found
-- Looking for CL_VERSION_1_0
-- Looking for CL_VERSION_1_0 - not found
-- Could NOT find OpenCL (missing:  OpenCL_LIBRARY OpenCL_INCLUDE_DIR) 
-- Looking for pthread.h
-- Looking for pthread.h - found
-- Looking for pthread_create
-- Looking for pthread_create - not found
-- Looking for pthread_create in pthreads
-- Looking for pthread_create in pthreads - not found
-- Looking for pthread_create in pthread
-- Looking for pthread_create in pthread - found
-- Found Threads: TRUE  
-- Found CUDA: /usr/local/cuda (found suitable version "10.1", minimum required is "9.0") 
-- Found PkgConfig: /usr/bin/pkg-config (found version "0.29.1") 
-- Found cuDNN: /usr/lib/x86_64-linux-gnu/libcudnn.so.7 (found suitable version "7.6", minimum required is "4.0") 
-- Could NOT find OpenCL (missing:  OpenCL_LIBRARY OpenCL_INCLUDE_DIR) (Required is at least version "1.2")
-- Found OpenGL: /usr/lib/x86_64-linux-gnu/libGL.so  
-- Found FreeImage: /usr/include  
-- Checking for module 'fftw3'
--   Found fftw3, version 3.3.4
-- Found FFTW: /usr/include  
-- Checking for module 'cblas'
--   No package 'cblas' found
-- Looking for sys/types.h
-- Looking for sys/types.h - found
-- Looking for stdint.h
-- Looking for stdint.h - found
-- Looking for stddef.h
-- Looking for stddef.h - found
-- Check size of void*
-- Check size of void* - done
-- Checking for [Accelerate]
-- Checking for [vecLib]
-- Checking for [cblas - atlas]
-- Includes found
-- Looking for cblas_dgemm
-- Looking for cblas_dgemm - found
-- CBLAS Symbols FOUND
-- CBLAS library found
-- Found LAPACKE: /usr/lib/liblapacke.so;/usr/lib/openblas-base/liblapack.so;/usr/lib/openblas-base/libblas.so  
-- Found LAPACK: /usr/include  
-- Found Doxygen: /usr/bin/doxygen (found version "1.8.11") 
-- Check size of int
-- Check size of int - done
-- MKL: Thread Layer(Intel OpenMP) Interface(4-byte Integer)
-- Could NOT find MKL: Source the compilervars.sh or mklvars.sh scripts included with your installation of MKL. This script searches for the libraries in MKLROOT, LIBRARY_PATHS(Linux), and LIB(Windows) environment variables (missing:  MKL_INCLUDE_DIR MKL_Core_LINK_LIBRARY MKL_Interface_LINK_LIBRARY MKL_ThreadLayer_LINK_LIBRARY) 
-- Could NOT find MKL: Source the compilervars.sh or mklvars.sh scripts included with your installation of MKL. This script searches for the libraries in MKLROOT, LIBRARY_PATHS(Linux), and LIB(Windows) environment variables (missing:  MKL_INCLUDE_DIR MKL_Core_STATIC_LINK_LIBRARY MKL_Interface_STATIC_LINK_LIBRARY MKL_ThreadLayer_STATIC_LINK_LIBRARY) 
-- Boost version: 1.70.0
-- CUDA_architecture_build_targets: 7.5 
-- Performing Test group_flags
-- Performing Test group_flags - Success
-- Performing Test has_ignored_attributes_flag
-- Performing Test has_ignored_attributes_flag - Success
-- Performing Test has_all_warnings_flag
-- Performing Test has_all_warnings_flag - Success
-- Found PythonInterp: /usr/bin/python (found version "2.7.12") 
-- Check if compiler accepts -pthread
-- Check if compiler accepts -pthread - yes
-- Performing Test pthread_flag
-- Performing Test pthread_flag - Success
-- Configuring done
-- Generating done
-- Build files have been written to: /home/turgut/Documents/arrayfire/build
turgut@turgut:~/Documents/arrayfire/build$ make install
Scanning dependencies of target bin2cpp
[  0%] Building CXX object CMakeFiles/bin2cpp.dir/CMakeModules/bin2cpp.cpp.o
[  0%] Linking CXX executable bin2cpp
[  0%] Built target bin2cpp
Scanning dependencies of target cuda_nvrtc_kernel_headers_bin_target
[  1%] Compiling /usr/local/cuda/include/cuda_fp16.h to C++ source
[  1%] Compiling /usr/local/cuda/include/cuda_fp16.hpp to C++ source
[  1%] Compiling /usr/local/cuda/include/cuComplex.h to C++ source
[  1%] Compiling /usr/local/cuda/include/math_constants.h to C++ source
[  1%] Compiling /home/turgut/Documents/arrayfire/src/api/c/optypes.hpp to C++ source
[  1%] Compiling /home/turgut/Documents/arrayfire/include/af/defines.h to C++ source
[  1%] Compiling /home/turgut/Documents/arrayfire/include/af/traits.hpp to C++ source
[  1%] Compiling /home/turgut/Documents/arrayfire/build/include/af/version.h to C++ source
[  1%] Compiling /home/turgut/Documents/arrayfire/src/backend/cuda/Param.hpp to C++ source
[  1%] Compiling /home/turgut/Documents/arrayfire/src/backend/cuda/assign_kernel_param.hpp to C++ source
[  1%] Compiling /home/turgut/Documents/arrayfire/src/backend/cuda/backend.hpp to C++ source
[  1%] Compiling /home/turgut/Documents/arrayfire/src/backend/cuda/dims_param.hpp to C++ source
[  2%] Compiling /home/turgut/Documents/arrayfire/src/backend/cuda/kernel/interp.hpp to C++ source
[  2%] Compiling /home/turgut/Documents/arrayfire/src/backend/cuda/kernel/shared.hpp to C++ source
[  2%] Compiling /home/turgut/Documents/arrayfire/src/backend/cuda/math.hpp to C++ source
[  2%] Compiling /home/turgut/Documents/arrayfire/src/backend/cuda/minmax_op.hpp to C++ source
[  2%] Compiling /home/turgut/Documents/arrayfire/src/backend/cuda/utility.hpp to C++ source
[  2%] Compiling /home/turgut/Documents/arrayfire/src/backend/cuda/types.hpp to C++ source
[  2%] Compiling /home/turgut/Documents/arrayfire/src/backend/cuda/../common/Binary.hpp to C++ source
[  2%] Compiling /home/turgut/Documents/arrayfire/src/backend/cuda/../common/Transform.hpp to C++ source
[  2%] Compiling /home/turgut/Documents/arrayfire/src/backend/cuda/../common/half.hpp to C++ source
[  2%] Compiling /home/turgut/Documents/arrayfire/src/backend/cuda/../common/internal_enums.hpp to C++ source
[  2%] Compiling /home/turgut/Documents/arrayfire/src/backend/cuda/../common/kernel_type.hpp to C++ source
[  2%] Compiling /home/turgut/Documents/arrayfire/src/backend/cuda/kernel/anisotropic_diffusion.cuh to C++ source
[  3%] Compiling /home/turgut/Documents/arrayfire/src/backend/cuda/kernel/approx1.cuh to C++ source
[  3%] Compiling /home/turgut/Documents/arrayfire/src/backend/cuda/kernel/approx2.cuh to C++ source
[  3%] Compiling /home/turgut/Documents/arrayfire/src/backend/cuda/kernel/assign.cuh to C++ source
[  3%] Compiling /home/turgut/Documents/arrayfire/src/backend/cuda/kernel/bilateral.cuh to C++ source
[  3%] Compiling /home/turgut/Documents/arrayfire/src/backend/cuda/kernel/canny.cuh to C++ source
[  3%] Compiling /home/turgut/Documents/arrayfire/src/backend/cuda/kernel/convolve1.cuh to C++ source
[  3%] Compiling /home/turgut/Documents/arrayfire/src/backend/cuda/kernel/convolve2.cuh to C++ source
[  3%] Compiling /home/turgut/Documents/arrayfire/src/backend/cuda/kernel/convolve3.cuh to C++ source
[  3%] Compiling /home/turgut/Documents/arrayfire/src/backend/cuda/kernel/convolve_separable.cuh to C++ source
[  3%] Compiling /home/turgut/Documents/arrayfire/src/backend/cuda/kernel/copy.cuh to C++ source
[  3%] Compiling /home/turgut/Documents/arrayfire/src/backend/cuda/kernel/diagonal.cuh to C++ source
[  3%] Compiling /home/turgut/Documents/arrayfire/src/backend/cuda/kernel/diff.cuh to C++ source
[  3%] Compiling /home/turgut/Documents/arrayfire/src/backend/cuda/kernel/exampleFunction.cuh to C++ source
[  4%] Compiling /home/turgut/Documents/arrayfire/src/backend/cuda/kernel/fftconvolve.cuh to C++ source
[  4%] Compiling /home/turgut/Documents/arrayfire/src/backend/cuda/kernel/flood_fill.cuh to C++ source
[  4%] Compiling /home/turgut/Documents/arrayfire/src/backend/cuda/kernel/gradient.cuh to C++ source
[  4%] Compiling /home/turgut/Documents/arrayfire/src/backend/cuda/kernel/histogram.cuh to C++ source
[  4%] Compiling /home/turgut/Documents/arrayfire/src/backend/cuda/kernel/hsv_rgb.cuh to C++ source
[  4%] Compiling /home/turgut/Documents/arrayfire/src/backend/cuda/kernel/identity.cuh to C++ source
[  4%] Compiling /home/turgut/Documents/arrayfire/src/backend/cuda/kernel/iir.cuh to C++ source
[  4%] Compiling /home/turgut/Documents/arrayfire/src/backend/cuda/kernel/index.cuh to C++ source
[  4%] Compiling /home/turgut/Documents/arrayfire/src/backend/cuda/kernel/iota.cuh to C++ source
[  4%] Compiling /home/turgut/Documents/arrayfire/src/backend/cuda/kernel/ireduce.cuh to C++ source
[  4%] Compiling /home/turgut/Documents/arrayfire/src/backend/cuda/kernel/join.cuh to C++ source
[  4%] Compiling /home/turgut/Documents/arrayfire/src/backend/cuda/kernel/lookup.cuh to C++ source
[  5%] Compiling /home/turgut/Documents/arrayfire/src/backend/cuda/kernel/lu_split.cuh to C++ source
[  5%] Compiling /home/turgut/Documents/arrayfire/src/backend/cuda/kernel/match_template.cuh to C++ source
[  5%] Compiling /home/turgut/Documents/arrayfire/src/backend/cuda/kernel/meanshift.cuh to C++ source
[  5%] Compiling /home/turgut/Documents/arrayfire/src/backend/cuda/kernel/medfilt.cuh to C++ source
[  5%] Compiling /home/turgut/Documents/arrayfire/src/backend/cuda/kernel/memcopy.cuh to C++ source
[  5%] Compiling /home/turgut/Documents/arrayfire/src/backend/cuda/kernel/moments.cuh to C++ source
[  5%] Compiling /home/turgut/Documents/arrayfire/src/backend/cuda/kernel/morph.cuh to C++ source
[  5%] Compiling /home/turgut/Documents/arrayfire/src/backend/cuda/kernel/pad_array_borders.cuh to C++ source
[  5%] Compiling /home/turgut/Documents/arrayfire/src/backend/cuda/kernel/range.cuh to C++ source
[  5%] Compiling /home/turgut/Documents/arrayfire/src/backend/cuda/kernel/resize.cuh to C++ source
[  5%] Compiling /home/turgut/Documents/arrayfire/src/backend/cuda/kernel/reorder.cuh to C++ source
[  5%] Compiling /home/turgut/Documents/arrayfire/src/backend/cuda/kernel/rotate.cuh to C++ source
[  5%] Compiling /home/turgut/Documents/arrayfire/src/backend/cuda/kernel/select.cuh to C++ source
[  6%] Compiling /home/turgut/Documents/arrayfire/src/backend/cuda/kernel/scan_dim.cuh to C++ source
[  6%] Compiling /home/turgut/Documents/arrayfire/src/backend/cuda/kernel/scan_dim_by_key.cuh to C++ source
[  6%] Compiling /home/turgut/Documents/arrayfire/src/backend/cuda/kernel/scan_first.cuh to C++ source
[  6%] Compiling /home/turgut/Documents/arrayfire/src/backend/cuda/kernel/scan_first_by_key.cuh to C++ source
[  6%] Compiling /home/turgut/Documents/arrayfire/src/backend/cuda/kernel/sobel.cuh to C++ source
[  6%] Compiling /home/turgut/Documents/arrayfire/src/backend/cuda/kernel/sparse.cuh to C++ source
[  6%] Compiling /home/turgut/Documents/arrayfire/src/backend/cuda/kernel/sparse_arith.cuh to C++ source
[  6%] Compiling /home/turgut/Documents/arrayfire/src/backend/cuda/kernel/susan.cuh to C++ source
[  6%] Compiling /home/turgut/Documents/arrayfire/src/backend/cuda/kernel/tile.cuh to C++ source
[  6%] Compiling /home/turgut/Documents/arrayfire/src/backend/cuda/kernel/transform.cuh to C++ source
[  6%] Compiling /home/turgut/Documents/arrayfire/src/backend/cuda/kernel/transpose.cuh to C++ source
[  6%] Compiling /home/turgut/Documents/arrayfire/src/backend/cuda/kernel/transpose_inplace.cuh to C++ source
[  7%] Compiling /home/turgut/Documents/arrayfire/src/backend/cuda/kernel/triangle.cuh to C++ source
[  7%] Compiling /home/turgut/Documents/arrayfire/src/backend/cuda/kernel/unwrap.cuh to C++ source
[  7%] Compiling /home/turgut/Documents/arrayfire/src/backend/cuda/kernel/where.cuh to C++ source
[  7%] Compiling /home/turgut/Documents/arrayfire/src/backend/cuda/kernel/wrap.cuh to C++ source
[  7%] Built target cuda_nvrtc_kernel_headers_bin_target
[  8%] Building NVCC (Device) object src/backend/cuda/CMakeFiles/af_cuda_static_cuda_library.dir/af_cuda_static_cuda_library_generated_solve.cu.o
[  8%] Building NVCC (Device) object src/backend/cuda/CMakeFiles/af_cuda_static_cuda_library.dir/af_cuda_static_cuda_library_generated_blas.cu.o
/home/turgut/Documents/arrayfire/src/backend/common/jit/NaryNode.hpp(43): error: static assertion failed with "NaryNode is not move assignable"

/home/turgut/Documents/arrayfire/src/backend/common/jit/UnaryNode.hpp(19): error: static assertion failed with "UnaryNode is not move assignable"

2 errors detected in the compilation of "/tmp/tmpxft_00003bfe_00000000-6_blas.cpp1.ii".
CMake Error at af_cuda_static_cuda_library_generated_blas.cu.o.cmake:266 (message):
  Error generating file
  /home/turgut/Documents/arrayfire/build/src/backend/cuda/CMakeFiles/af_cuda_static_cuda_library.dir//./af_cuda_static_cuda_library_generated_blas.cu.o


src/backend/cuda/CMakeFiles/af_cuda_static_cuda_library.dir/build.make:63: recipe for target 'src/backend/cuda/CMakeFiles/af_cuda_static_cuda_library.dir/af_cuda_static_cuda_library_generated_blas.cu.o' failed
make[2]: *** [src/backend/cuda/CMakeFiles/af_cuda_static_cuda_library.dir/af_cuda_static_cuda_library_generated_blas.cu.o] Error 1
CMakeFiles/Makefile2:1204: recipe for target 'src/backend/cuda/CMakeFiles/af_cuda_static_cuda_library.dir/all' failed
make[1]: *** [src/backend/cuda/CMakeFiles/af_cuda_static_cuda_library.dir/all] Error 2
Makefile:160: recipe for target 'all' failed
make: *** [all] Error 2
turgut@turgut:~/Documents/arrayfire/build$ 

```

Build Environment
-----------------
```gcc (Ubuntu 5.4.0-6ubuntu1~16.04.12) 5.4.0 20160609```

--- a/docs/pages/release_notes.md
+++ b/docs/pages/release_notes.md
@@ -1,6 +1,96 @@
 Release Notes {#releasenotes}
 ==============
 
+v3.8.0
+======
+
+Major Updates
+-------------
+- Ragged max
+- Bitwise not
+- Updated alloc and free
+- Initializer list for af::array
+
+Improvements
+------------
+
+v3.7.2
+======
+
+Improvements
+------------
+- Cache CUDA kernels to disk to improve load times(Thanks to \@cschreib-ibex) /PR{2848}
+- Staticly link against cuda libraries /PR{2785}
+- Make cuDNN an optional build dependency /PR{2836}
+- Improve support for different compilers and OS /PR{2876} /PR{2945} /PR{2925} /PR{2942} /PR{2943} /PR{2945}
+- Improve performance of join and transpose on CPU /PR{2849}
+- Improve documentation /PR{2816} /PR{2821} /PR{2846} /PR{2918} /PR{2928} /PR{2947}
+- Reduce binary size using NVRTC and template reducing instantiations /PR{2849} /PR{2861} /PR{2890}
+- Improve reduceByKey performance on OpenCL by using builtin functions /PR{2851}
+- Improve support for Intel OpenCL GPUs /PR{2855}
+- Allow staticly linking against MKL /PR{2877} (Sponsered by SDL)
+- Better support for older CUDA toolkits /PR{2923}
+- Add support for CUDA 11 /PR{2939}
+- Add support for ccache for faster builds /PR{2931}
+- Add support for the conan package manager on linux /PR{2875}
+
+Fixes
+-----
+- Bug crash when allocating large arrays /PR{2827}
+- Fix various compiler warnings /PR{2827} /PR{2849} /PR{2872} /PR{2876}
+- Fix minor leaks in OpenCL functions /PR{2913}
+- Various continuous integration related fixes /PR{2819}
+- Fix zero padding with convolv2NN /PR{2820}
+- Fix af_get_memory_pressure_threshold return value /PR{2831}
+- Increased the max filter length for morph
+- Handle empty array inputs for LU, QR, and Rank functions /PR{2838}
+- Fix FindMKL.cmake script for sequential threading library /PR{2840}
+- Various internal refactoring /PR{2839} /PR{2861} /PR{2864} /PR{2873} /PR{2890} /PR{2891} /PR{2913}
+- Fix OpenCL 2.0 builtin function name conflict /PR{2851}
+- Fix error caused when releasing memory with multiple devices /PR{2867}
+
+Contributions
+-------------
+Special thanks to our contributors:
+[Corentin Schreiber](https://github.com/cschreib-ibex)
+[Jacob Khan](https://github.com/jacobkahn)
+[Paul Jurczak](https://github.com/pauljurczak)
+
+v3.7.1
+======
+
+Improvements
+------------
+
+- Improve mtx download for test data \PR{2742}
+- Documentation improvements \PR{2754} \PR{2792} \PR{2797}
+- Remove verbose messages in older CMake versions \PR{2773}
+- Reduce binary size with the use of nvrtc  \PR{2790}
+- Use texture memory to load LUT in orb and fast \PR{2791}
+- Add missing print function for f16 \PR{2784}
+- Add checks for f16 support in the CUDA backend \PR{2784}
+- Create a thrust policy to intercept tmp buffer allocations \PR{2806}
+
+Fixes
+-----
+
+- Fix segfault on exit when ArrayFire is not initialized in the main thread
+- Fix support for CMake 3.5.1 \PR{2771} \PR{2772} \PR{2760}
+- Fix evalMultiple if the input array sizes aren't the same \PR{2766}
+- Fix error when AF_BACKEND_DEFAULT is passed directly to backend \PR{2769}
+- Workaround name collision with AMD OpenCL implementation \PR{2802}
+- Fix on-exit errors with the unified backend \PR{2769}
+- Fix check for f16 compatibility in OpenCL \PR{2773}
+- Fix matmul on Intel OpenCL when passing same array as input \PR{2774}
+- Fix CPU OpenCL blas batching \PR{2774}
+- Fix memory pressure in the default memory manager \PR{2801}
+
+Contributions
+-------------
+Special thanks to our contributors:
+[padentomasello](https://github.com/padentomasello)
+[glavaux2](https://github.com/glavaux2)
+
 v3.7.0
 ======
 
@@ -205,7 +295,7 @@ Misc
 Contributions
 -------------
 Special thanks to our contributors: [Jacob Kahn](https://github.com/jacobkahn),
-[Vardan Akopian](https://github.com/vakopian)  
+[Vardan Akopian](https://github.com/vakopian)
 
 v3.6.1
 ======

--- a/src/backend/common/half.hpp
+++ b/src/backend/common/half.hpp
@@ -832,7 +832,10 @@ class alignas(2) half {
 #endif
 
    public:
-    AF_CONSTEXPR half() = default;
+#if CUDA_VERSION >= 10000
+    AF_CONSTEXPR
+#endif
+    half() = default;
 
     /// Constructor.
     /// \param bits binary representation to set half to

--- a/src/backend/common/jit/NaryNode.hpp
+++ b/src/backend/common/jit/NaryNode.hpp
@@ -26,7 +26,7 @@ class NaryNode : public Node {
    private:
     int m_num_children;
     int m_op;
-    std::string m_op_str;
+    const char *m_op_str;
 
    public:
     NaryNode(const af::dtype type, const char *op_str, const int num_children,
@@ -46,7 +46,7 @@ class NaryNode : public Node {
                       "NaryNode is not move constructible");
     }
 
-    NaryNode(NaryNode &&other) = default;
+    NaryNode(NaryNode &&other) noexcept = default;
 
     NaryNode(const NaryNode &other) = default;
 

--- a/src/backend/common/jit/Node.hpp
+++ b/src/backend/common/jit/Node.hpp
@@ -106,11 +106,17 @@ class Node {
                       "Node is not move assignable");
     }
 
+    /// Default move constructor operator
+    Node(Node &&node) noexcept = default;
+
     /// Default copy constructor operator
     Node(const Node &node) = default;
 
     /// Default copy assignment operator
     Node &operator=(const Node &node) = default;
+
+    /// Default move assignment operator
+    Node &operator=(Node &&node) noexcept = default;
 
     int getNodesMap(Node_map_t &node_map, std::vector<Node *> &full_nodes,
                     std::vector<Node_ids> &full_ids);

--- a/src/backend/cuda/CMakeLists.txt
+++ b/src/backend/cuda/CMakeLists.txt
@@ -54,7 +54,10 @@ if(UNIX)
   # use ${CMAKE_*_LIBRARY} variables in the following flags.
   set(af_cuda_static_flags "${af_cuda_static_flags};-lculibos")
   set(af_cuda_static_flags "${af_cuda_static_flags};-lcublas_static")
-  set(af_cuda_static_flags "${af_cuda_static_flags};-lcublasLt_static")
+
+  if(CUDA_VERSION VERSION_GREATER 10.0)
+    set(af_cuda_static_flags "${af_cuda_static_flags};-lcublasLt_static")
+  endif()
   set(af_cuda_static_flags "${af_cuda_static_flags};-lcufft_static")
   set(af_cuda_static_flags "${af_cuda_static_flags};-lcusparse_static")
 
@@ -323,10 +326,14 @@ if(UNIX)
       ${END_GROUP}
   )
 
+  if(CUDA_VERSION VERSION_GREATER 10.0)
+    target_link_libraries(af_cuda_static_cuda_library
+      PRIVATE
+        ${CUDA_cublasLt_static_LIBRARY})
+  endif()
   if(CUDA_VERSION VERSION_GREATER 9.5)
     target_link_libraries(af_cuda_static_cuda_library
       PRIVATE
-        ${CUDA_cublasLt_static_LIBRARY}
         ${CUDA_lapack_static_LIBRARY})
   endif()
 
@@ -798,7 +805,9 @@ if(AF_INSTALL_STANDALONE)
   if(WIN32)
     afcu_collect_libs(cufft)
     afcu_collect_libs(cublas)
-    afcu_collect_libs(cublasLt)
+    if(CUDA_VERSION VERSION_GREATER 10.0)
+      afcu_collect_libs(cublasLt)
+    endif()
     afcu_collect_libs(cusolver)
     afcu_collect_libs(cusparse)
   elseif(NOT ${use_static_cuda_lapack})

--- a/src/backend/cuda/kernel/random_engine.hpp
+++ b/src/backend/cuda/kernel/random_engine.hpp
@@ -10,6 +10,7 @@
 #pragma once
 
 #include <common/dispatch.hpp>
+#include <common/half.hpp>
 #include <debug_cuda.hpp>
 #include <err_cuda.hpp>
 #include <kernel/random_engine_mersenne.hpp>
@@ -597,7 +598,7 @@ __device__ static void partialWriteOut128Bytes(common::half *out,
 __device__ static void partialBoxMullerWriteOut128Bytes(
     common::half *out, const uint &index, const uint &r1, const uint &r2,
     const uint &r3, const uint &r4, const uint &elements) {
-    common::half n[8];
+    __half n[8];
     boxMullerTransform(n + 0, n + 1, getHalf(r1), getHalf(r1 >> 16));
     boxMullerTransform(n + 2, n + 3, getHalf(r2), getHalf(r2 >> 16));
     boxMullerTransform(n + 4, n + 5, getHalf(r3), getHalf(r3 >> 16));

--- a/src/backend/cuda/kernel/reduce_by_key.hpp
+++ b/src/backend/cuda/kernel/reduce_by_key.hpp
@@ -106,9 +106,6 @@ __global__ void compact(int *reduced_block_sizes, Param<Tk> keys_out,
     const int bidz = blockIdx.z % nBlocksZ;
     const int bidw = blockIdx.z / nBlocksZ;
 
-    Tk k;
-    To v;
-
     // reduced_block_sizes should have inclusive sum of block sizes
     int nwrite = (blockIdx.x == 0) ? reduced_block_sizes[0]
                                    : reduced_block_sizes[blockIdx.x] -
@@ -117,8 +114,8 @@ __global__ void compact(int *reduced_block_sizes, Param<Tk> keys_out,
 
     const int bOffset = bidw * vals_in.strides[3] + bidz * vals_in.strides[2] +
                         bidy * vals_in.strides[1];
-    k = keys_in.ptr[tidx];
-    v = vals_in.ptr[bOffset + tidx];
+    Tk k = keys_in.ptr[tidx];
+    To v = vals_in.ptr[bOffset + tidx];
 
     if (threadIdx.x < nwrite) {
         keys_out.ptr[writeloc + threadIdx.x]           = k;
@@ -147,9 +144,6 @@ __global__ void compact_dim(int *reduced_block_sizes, Param<Tk> keys_out,
     const int bidz = blockIdx.z % nBlocksZ;
     const int bidw = blockIdx.z / nBlocksZ;
 
-    Tk k;
-    To v;
-
     // reduced_block_sizes should have inclusive sum of block sizes
     int nwrite = (blockIdx.x == 0) ? reduced_block_sizes[0]
                                    : reduced_block_sizes[blockIdx.x] -
@@ -160,8 +154,8 @@ __global__ void compact_dim(int *reduced_block_sizes, Param<Tk> keys_out,
                     bidz * vals_in.strides[dim_ordering[2]] +
                     bidy * vals_in.strides[dim_ordering[1]] +
                     tidx * vals_in.strides[dim];
-    k = keys_in.ptr[tidx];
-    v = vals_in.ptr[tid];
+    Tk k = keys_in.ptr[tidx];
+    To v = vals_in.ptr[tid];
 
     if (threadIdx.x < nwrite) {
         keys_out.ptr[writeloc + threadIdx.x] = k;

